engine/classes/Elgg/UserCapabilities.php

```diff
@@ -0,0 +1,301 @@
+			return false;
+		}
+
+		if ($user) {
+			$user_guid = $user->guid;
+		}
+
+		$params = [
+			'container' => $entity,
+			'user' => $user,
+			'subtype' => $subtype
+		];
+
+		// Unlike permissions, logic check can be used to prevent certain entity
+		// types from being contained by other entity types,
+		// e.g. discussion reply objects can only be contained by discussion objects.
+		// This hook can also be used to apply status logic, e.g. to disallow
+		// new replies in closed discussions.
+		// We do not take a stand hence the return is null. This can be used by
+		// handlers to check if another hook has modified the value.
+		$logic_check = $this->hooks->trigger('container_logic_check', $type, $params);
+
+		if ($logic_check === false) {
+			return false;
+		}
+		
+		$return = false;
+		if ($entity) {
+			// If the user can edit the container, they can also write to it
+			if ($entity->canEdit($user_guid)) {
+				$return = true;
+			}
+		}
+
+		// Container permissions can prevent users from writing to an entity.
+		// For instance, these permissions can prevent non-group members from writing
+		// content to the group.
+		return $this->hooks->trigger('container_permissions_check', $type, $params, $return);
+	}
+
+	/**
+	 * Can a user comment on an entity?
+	 *
+	 * @tip Can be overridden by registering for the permissions_check:comment,
+	 * <entity type> plugin hook.
+	 *
+	 * @param ElggEntity $entity    Object entity
+	 * @param int        $user_guid User guid (default is logged in user)
+	 *
+	 * @return bool
+	 */
+	public function canComment(ElggEntity $entity, $user_guid = 0) {
+		try {
+			$user = $this->entities->getUserForPermissionsCheck($user_guid);
+		} catch (UserFetchFailureException $e) {
+			return false;
+		}
+
+		// By default, we don't take a position of whether commenting is allowed
+		// because it is handled by the subclasses of \ElggEntity
+		$params = [
+			'entity' => $entity,
+			'user' => $user
+		];
+		return $this->hooks->trigger('permissions_check:comment', $entity->getType(), $params, null);
+	}
+
+	/**
+	 * Can a user annotate an entity?
+	 *
+	 * @tip Can be overridden by registering for the plugin hook [permissions_check:annotate:<name>,
+	 * <entity type>] or [permissions_check:annotate, <entity type>]. The hooks are called in that order.
+	 *
+	 * @tip If you want logged out users to annotate an object, do not call
+	 * canAnnotate(). It's easier than using the plugin hook.
+	 *
+	 * @param ElggEntity $entity          Objet entity
+	 * @param int        $user_guid       User guid (default is logged in user)
+	 * @param string     $annotation_name The name of the annotation (default is unspecified)
+	 *
+	 * @return bool
+	 */
+	public function canAnnotate(ElggEntity $entity, $user_guid = 0, $annotation_name = '') {
+		if (!empty($annotation_name) && !is_string($annotation_name)) {
+			throw new \InvalidArgumentException(__METHOD__ . ' expects \$annotation_name to be a string');
```

Comment:
This would still accept an array. Do we really need to accept `null` and `false`?